ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/OrcFileMerger.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.ql.exec.vector.VectorizedRowBatch;
+import org.apache.hadoop.hive.ql.io.orc.CompressionKind;
+import org.apache.hadoop.hive.ql.io.orc.OrcFile;
+import org.apache.hadoop.hive.ql.io.orc.Reader;
+import org.apache.hadoop.hive.ql.io.orc.RecordReader;
+import org.apache.hadoop.hive.ql.io.orc.Writer;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Class to support fast merging of ORC files.
+ */
+public class OrcFileMerger {
+
+  private final Configuration conf;
+  private static final Logger LOG = LoggerFactory.getLogger(OrcFileMerger.class);
+
+  public OrcFileMerger(Configuration conf) {
+    this.conf = conf;
+  }
+
+  /**
+   * Merge orc files into a single file
+   * @param files list of orc file paths to be merged
+   * @param outPath the path of output orc file
+   * @throws IOException error happened during file operations
+   */
+  public void mergeFiles(List<Reader> readers, Path outPath) throws IOException {
+    Writer writer = null;
+    try {
+      for (Reader reader : readers) {
+        if (writer == null) {
+          writer = setupWriter(reader, outPath);
+        }
+        VectorizedRowBatch batch = reader.getSchema().createRowBatchV2();
+        RecordReader rows = reader.rows();
+        while (rows.nextBatch(batch)) {
+          if (batch != null) {
+            writer.addRowBatch(batch);
+          }
+        }
+        rows.close();
+      }
+    } finally {
+      if (writer != null) {
+        writer.close();
+      }
+    }
```

Comment:
If we create a writer irrespective of number of readers in try with resources block then a writer will always be created and this may create a empty file (if readers are not present).

Reply:
Ok, makes sense